src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/IntersectingPointValues.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1.bbtree;
+
+import java.io.IOException;
+
+import org.apache.lucene.codecs.MutablePointValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.util.BytesRef;
+
+public abstract class IntersectingPointValues extends MutablePointValues
+{
+    private static final byte[] EMPTY = new byte[0];
+
+    public boolean needsSorting()
+    {
+        return size() > 1;
+    }
+
+    public abstract void intersect(IntersectVisitor visitor) throws IOException;
+
+    @Override
+    public int getDocCount()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long size()
+    {
+        return 1;
+    }
+
+    @Override
+    public void getValue(int i, BytesRef packedValue)
+    {
+        // no-op
+    }
+
+    @Override
+    public byte getByteAt(int i, int k)
+    {
+        return 0;
```

Comment:
I've changed all the methods that are returning dummy values to return `UnsupportedOperationException`. These methods are only used in the sort operation and that is only done if `size() > 0`. This is only the case for the implementation in `BlockBalancedTreeRamBuffer` where the values need sorting.

On a related subject. It seems to me that the only reason that we are using `PointValues` at all is because of the sort. This is needed because the values in `BlockBalancedTreeRamBuffer` are not sorted. One option here would be to change how we build the in-memory numerics such that they are naturally sorted. I am thinking of a trie here. This would also give us prefix compression of the values as they are built in-memory, which we don't have with the current implementation.

Reply:
> One option here would be to change how we build the in-memory numerics such that they are naturally sorted. I am thinking of a trie here.

I like this, and I like it a lot. I'm +100 on just replacing `BlockBalancedTreeRamBuffer` entirely w/ an in-memory trie and blowing away all the `PointValues` garbage. Then we end up having something more like `ImmutableIntersectingPointValues` for all segment building cases?

What do we lose if we toss out `BlockBalancedTreeRamBuffer`? `PackedLongValues` packing row IDs more efficiently than the trie stores them as payloads? If losing that and gaining value prefix compression was about break-even, being able to avoid sorting would still be a win...

Reply:
We could look at packing the values for trie postings but even if we didn't I still think we would gain more from the trie prefix compression than we would lose from the `PackedLongValues`. I'm also looking at the `RAMStringIndexer` here. We could replace that with exactly the same in-memory structure. Whether it gave us much of an improvement or not, it would reduce code complexity and reduce our reliance on Lucene code.